chunkset.c

```diff
@@ -34,27 +34,24 @@ static inline void chunkmemset_1(uint8_t *from, chunk_t *chunk) {
 static inline void chunkmemset_4(uint8_t *from, chunk_t *chunk) {
 #if defined(UNALIGNED64_OK)
     uint32_t half_chunk;
-    half_chunk = *(uint32_t *)from;
+    memcpy(&half_chunk, from, sizeof(half_chunk));
     *chunk = 0x0000000100000001 * (uint64_t)half_chunk;
 #elif defined(UNALIGNED_OK)
-    chunk->u32[0] = *(uint32_t *)from;
+    memcpy(&chunk->u32[0], from, sizeof(chunk->u32[0]));
     chunk->u32[1] = chunk->u32[0];
 #else
     uint8_t *chunkptr = (uint8_t *)chunk;
-    memcpy(chunkptr, from, 4);
-    memcpy(chunkptr+4, from, 4);
+    memcpy(chunkptr, from, sizeof(uint32_t));
+    memcpy(chunkptr+4, from, sizeof(uint32_t));
```

Comment:
I think this second memcpy is erronous btw, I thought them using the same from pointer strange.
Removing the second one completely seems to make no difference.

Reply:
We are doing essentially a `memset` with size of 4. So our chunk is 8 bytes. We have to copy 4 byte value from `from` twice.

Reply:
No, I think this function is supposed to copy only 4 bytes.

Reply:
Yes, but the chunk size is 8 bytes. So it copies four bytes twice. If our chunk is `uint64_t` it needs to be able to use that `uint64_t` to perform multiple copies of the 4 bytes in a row.

Reply:
![image](https://user-images.githubusercontent.com/1364432/148619710-5863fabc-5435-4a7b-be48-5e8a189ca5c9.png)
That is the original code. You can see it only does 1 x 4 byte copy. But then it uses the instruction `_mm_set1_epi32` to duplicate that.

In our code here we are essentially doing what `_mm_set1_epi32` does but with 64-bit integer.

Reply:
I see your point now. I still don't grok the chunkset stuff it seems, I never quite did.

Reply:
It is easier to understand and look at one of the chunkset sources for instruction sets like `chunkset_sse.c`. There are still some parts of chunkset overall that are fuzzy to me as well.